substrate/frame/election-provider-support/src/lib.rs

```diff
@@ -361,72 +379,87 @@ pub trait ElectionDataProvider {
 	/// Clear all voters and targets.
 	#[cfg(any(feature = "runtime-benchmarks", test))]
 	fn clear() {}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	fn set_desired_targets(_count: u32) {}
 }
 
-/// Base trait for types that can provide election
-pub trait ElectionProviderBase {
-	/// The account identifier type.
+/// An [`ElectionDataProvider`] that exposes for an external entity to request a lock/unlock on
+/// updates in the election data.
+pub trait LockableElectionDataProvider: ElectionDataProvider {
+	fn set_lock() -> data_provider::Result<()>;
```

Comment:
Suggestion: Not sure how much change this is but I think the better election flow would be
- Staking asks `ElectionProvider` to start preparing for election. (right now this happens via system hooks)
- At this point staking knows EPM will call `ElectionDataProvider` to build the target and backer snapshot, and it shouldn't change until EPM is done. This is equivalent to set_lock().
- EPM builds the snapshot and signals somehow that it is done (or if last page is called, Staking knows its done). This would be equivalent to unlock().

Reply:
OTOH, I will play a bit with exposing how many blocks are needed for preparing an election in the EPM side and maybe simplify the whole phase transition calculation.

Reply:
I think for the objectives of this PR, changing the initiator of the overall interaction is too much change.